src/coreclr/vm/methodtablebuilder.cpp

```diff
@@ -8481,21 +8480,22 @@ MethodTableBuilder::HandleExplicitLayout(
         // "i" indexes all fields, valueClassCacheIndex indexes non-static fields only. Don't get them confused!
         valueClassCacheIndex++;
 
-        if (CorTypeInfo::IsObjRef(pFD->GetFieldType()))
+        CorElementType type = pFD->GetFieldType();
+        if (CorTypeInfo::IsObjRef(type) || CorTypeInfo::IsByRef(type))
         {
             // Check that the ref offset is pointer aligned
             if ((pFD->GetOffset_NoLogging() & ((ULONG)TARGET_POINTER_SIZE - 1)) != 0)
             {
                 badOffset = pFD->GetOffset_NoLogging();
                 fieldTrust.SetTrust(ExplicitFieldTrust::kNone);
 
-                // If we got here, OREF field was not pointer aligned. THROW.
+                // If we got here, OREF or BYREF field was not pointer aligned. THROW.
                 break;
             }
             // check if overlaps another object
             if (memcmp((void *)&pFieldLayout[pFD->GetOffset_NoLogging()], (void *)isObject, sizeof(isObject)) == 0)
             {
-                // If we got here, an OREF overlapped another OREF. We permit this but mark the class unverifiable.
+                // If we got here, an OREF or BYREF overlapped another OREF or BYREF. We permit this but mark the class unverifiable.
```

Comment:
Do we permit this?

```csharp
[Explicit]
ref struct Foo
{
    [FieldOffset(0)] object o;
    [FieldOffset(0)] ref int x;
}
```

I think this has the same problem as:

```csharp
[Explicit]
ref struct Foo
{
    [FieldOffset(0)] object o;
    [FieldOffset(0)] IntPtr x;
}
```

I think both have the same kinds of problems (setting the value through the other field to a non-object-pointer value is probably going to be bad when GC happens). If it was up to me, I wouldn't allow that.

We don't care much if byref points to a garbage, but objref pointing to garbage is not good.

Reply:
```csharp
[Explicit]
ref struct Foo
{
    [FieldOffset(0)] object o;
    [FieldOffset(0)] ref int x;
}
```

Mono handles that, but CoreCLR doesn't. Let me see what that would take.

Reply:
@MichalStrehovsky Unless there is pushback here, I don't think the overlap of an `object` and `ref` field should be a concern. Both Mono and CoreCLR handle the `IntPtr` case, which I agree is a concern. The `object` and `ref` scenario should be considered "okay" because there should be no way in safe C# to create an invalid `ref`.

Reply:
Bah. I've been correct here. This is an issue.